### PR TITLE
Allow rendering <input> and <textarea> inside amp-mustache

### LINKED ISSUE #5207: Consider whitelisting <input> in Mustache template rendering
Currently `input` elements are blacklisted (and hence stripped out) when rendering mustache templates.

So the following would just render nothing when rendered.

```html
<template type="amp-mustache">
  <input type="text" name="name">
</template>
```

This was initially blacklisted as a conservative measure to make sure this wasn't abused. We need to research this and see if whitelisting it wouldn't cause any security-holes or abuse. If it's safe to whitelist, this is can be pretty useful in the `form` template rendering case!

cc/ @dvoytenko

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #5207

### CODE REVIEW

**Comment by molnarg** [comment]
LGTM

File: src/sanitizer.js [hunk 1]
```diff
@@ -66,6 +64,7 @@ const SELF_CLOSING_TAGS = {
   'col': true,
   'hr': true,
   'img': true,
+  'input': true,
```
Comment (dvoytenko) [inline src/sanitizer.js]: Are there any other I missed? I'd rather have a full list there, even if it's blacklisted tags.

File: src/sanitizer.js [hunk 2]
```diff
@@ -116,6 +115,37 @@ const BLACKLISTED_ATTR_VALUES = [
 ];
 
 
+/** @const {!Object<string, !Object<string, !RegExp>>} */
+const BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES = {
+  'input': {
+    'type': /(?:image|file|password|button)/i,
```
Comment (dvoytenko) [inline src/sanitizer.js]: It'd seem that we actually do allow `<button>` itself. Why prohibit it here?

File: src/sanitizer.js (hunk 1 above)
Comment (mkhatib) [inline src/sanitizer.js]: Will research and update this list.

File: src/sanitizer.js (hunk 2 above)
Comment (mkhatib) [inline src/sanitizer.js]: Input[type=button] is not allowed in validator so this is here to match that as well. I don't remember the reason why we blacklisted it but can try to look it up and see. <button> is still allowed.

Comment (dvoytenko) [inline src/sanitizer.js]: Please follow validator here then. But also please send an issue to validator folk about this.

File: src/sanitizer.js (hunk 1 above)
Comment (mkhatib) [inline src/sanitizer.js]: Updated

File: src/sanitizer.js (hunk 2 above)
Comment (mkhatib) [inline src/sanitizer.js]: You mean to allow `input[type=button]`?

File: src/sanitizer.js (hunk 1 above)
Comment (mkhatib) [inline src/sanitizer.js]: @dvoytenko adding `link` to self-closing tags causes a test to fail:

```bash
Chromium 51.0.2704 (Ubuntu 0.0.0) sanitizeHtml should NOT output security-sensitive markup FAILED
	AssertionError: expected 'a</link>c' to equal 'ac'
	    at Context.<anonymous> amphtml/test/functional/test-sanitizer.js:56:50)
```

This is fine right? We just need to fix the test to drop the closing `</link>` tag no?

Comment (dvoytenko) [inline src/sanitizer.js]: Yeah. That test needs to be updated.

**Comment by dvoytenko** [comment]
LGTM, with the request to cleanup the self-closing tags.

File: src/sanitizer.js (hunk 2 above)
Comment (mkhatib) [inline src/sanitizer.js]: Not that useful, I remember chatting with @cramforce about it and deciding to just allow input[type=submit] and button.

**Pull request merged**